src/mix_protocol.nim

```diff
@@ -49,123 +52,117 @@ proc cryptoRandomInt(max: int): Result[int, string] =
-        error "Deserialization failed", err = error
-        return
+    let unpaddedMsg = unpadMessage(msgChunk).valueOr:
+      error "Unpadding message failed", err = error
+      return
 
-      let (message, protocol) = getMixMessage(deserializedResult)
-      info "# Received: ", receiver = multiAddr, message = message
+    let deserializedResult = deserializeMixMessage(unpaddedMsg).valueOr:
+      error "Deserialization failed", err = error
+      return
 
-      # Add delay
-      let delayMillis = (delay[0].int shl 8) or delay[1].int
-      await sleepAsync(milliseconds(delayMillis))
+    let
+      (message, protocol) = getMixMessage(deserializedResult)
+      exitConn = MixExitConnection.new(message)
+    trace "# Received: ", receiver = multiAddr, message = message
+    await mixProto.pHandler(exitConn, protocol)
 
-      # Forward to destination
-      let destBytes = getHop(nextHop)
+    if exitConn != nil:
+      try:
+        await exitConn.close()
+      except CatchableError as e:
+        error "Failed to close exit connection: ", err = e.msg
+  of Success:
+    trace "# Intermediate: ", multiAddr = multiAddr
+    # Add delay
+    let delayMillis = (delay[0].int shl 8) or delay[1].int
+    await sleepAsync(milliseconds(delayMillis))
+
+    # Forward to next hop
+    let nextHopBytes = getHop(nextHop)
+
+    let fullAddrStr = bytesToMultiAddr(nextHopBytes).valueOr:
+      error "Failed to convert bytes to multiaddress", err = error
+      return
 
-      let fullAddrStr = bytesToMultiAddr(destBytes).valueOr:
-        error "Failed to convert bytes to multiaddress", err = error
-        return
+    let parts = fullAddrStr.split("/p2p/")
+    if parts.len != 2:
+      error "Invalid multiaddress format", parts = parts
+      return
 
-      let parts = fullAddrStr.split("/p2p/")
-      if parts.len != 2:
-        error "Invalid multiaddress format", parts = parts
-        return
+    let
+      locationAddrStr = parts[0]
+      peerIdStr = parts[1]
 
-      let
-        locationAddrStr = parts[0]
-        peerIdStr = parts[1]
+    # Create MultiAddress and PeerId
+    let locationAddr = MultiAddress.init(locationAddrStr).valueOr:
+      error "Failed to parse location multiaddress: ", err = error
+      return
 
-      # Create MultiAddress and PeerId
-      let locationAddr = MultiAddress.init(locationAddrStr).valueOr:
-        error "Failed to parse location multiaddress: ", err = error
-        return
+    let peerId = PeerId.init(peerIdStr).valueOr:
+      error "Failed to initialize PeerId", err = error
+      return
 
-      let peerId = PeerId.init(peerIdStr).valueOr:
-        error "Failed to initialize PeerId", err = error
-        return
-
-      var destConn: Connection
-      try:
-        destConn = await mixProto.switch.dial(peerId, @[locationAddr], $protocol)
-        await destConn.writeLp(message)
-      except CatchableError as e:
-        error "Failed to dial next hop: ", err = e.msg
```

Comment:
This will need some sort of logic branch to account for "exit is dest" vs "exit forwards to dest"